ug: add checkpoints tutorial (#2373)
iterative/dvc.org: content/docs/user-guide/experiment-management/checkpoints.md

want to track, which you'll see later in this tutorial.

## Setting up the project

This tutorial is going to cover how to implement checkpoints in a ML project
using DVC. We're going to train a model to identify handwritten digits based on
the MNIST dataset.

You can follow along with the steps here or you can clone the repo directly from
GitHub and play with it. To clone the repo, run the following commands.

```bash
git clone https://github.com/iterative/checkpoints-tutorial
cd checkpoints-tutorial
```

It is highly recommended you create a virtual environment for this example. You
can do that by running:

```bash
python3 -m venv .venv
```

Once your virtual environment is installed, you can start it up with one of the
following commands.

- On Mac/Linux: `source .venv/bin/activate`
- On Windows: `.\.venv\Scripts\activate`

Once you have your environment set up, you can install the dependencies by
running:

```bash
pip install -r requirements.txt
```

This will download all of the packages you need to run the example. Now you have
everything you need to get started with experiments and checkpoints.

## Setting up a DVC pipeline

> If you don't have a DVC pipeline setup in your project, adding one only takes a

Reviewer: It is still not clear why user needs a DVC pipeline. You can find it here: https://github.com/iterative/dvc.org/pull/2373#discussion_r622459438

few commands. At the root of the project, run:

```bash
dvc init
```

This sets up the files you need for your DVC pipeline to work.

This will create a _dvc.yaml_ file in your root directory.

Now we need to add a stage for training our model within a DVC pipeliene. We'll
do that with `dvc stage add`, which we'll explain more later. For now, run the
following command:

```bash

> dvc stage add -n train -d data/MNIST -d train.py -c model.pt --plots-no-cache predictions.json -p seed,lr,weight_decay --live dvclive python train.py

Reviewer: New lines are needed like:
```bash
dvc stage add -n train -d data/MNIST -d train.py \
                             -c model.pt --plots-no-cache predictions.json
                             -p seed,lr,weight_decay --live dvclive python train.py
```

```

Before we go any further, this is a great point to add these changes to your Git
history. You can do that with the following commands:

```bash
git add .
git commit -m "created DVC pipeline"
```

> ## Enabling DVC pipelines for checkpoint experiments

Reviewer: I'd suggest moving it right after the command run `dvc stages add` and before Git commit. Otherwise, it looks for a user as a separate action.

The explanation can be reduced a bit.

setting up checkpoints in your code.

## Registering checkpoints in your code

Take a look at the _train.py_ file and you'll see how we train a convolutional
neural network to classify handwritten digits. The main area of this code most
relevant to checkpoints is when we iterate over the training epochs.

This is where DVC will be able to track our metrics over time and where we will
add our checkpoints to give us the points in time with our model that we can
switch between.

Now we need to enable checkpoints at the code level. We are interested in
tracking the metrics along with each checkpoint, so we'll need to add a few
lines of code.

In the `train.py` file, import the `dvclive` package with the other imports:

```bash
import dvclive
```

Then update the following lines of code in the `main` method inside of the
training epoch loop.

```python
for k, v in metrics.items():
    print('Epoch %s: %s=%s'%(i, k, v))
    dvclive.log(k, v)
dvclive.next_step()
```

By adding the two `dvclive` methods to our training epoch loop, we are enabling
checkpoints in our code and recording the training metrics at each of those

> checkpoints.

Reviewer: A bit more details might be helpful: https://github.com/iterative/dvc.org/pull/2373#discussion_r622482174

│ │ ╟ d50c724           │ 05:03 PM │    6 │ 0.33246 │ 0.9044 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ 29491a9           │ 05:02 PM │    5 │ 0.36601 │ 0.8943 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ b3de55f           │ 05:02 PM │    4 │ 0.41666 │ 0.8777 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ c4a46af           │ 05:02 PM │    3 │ 0.50835 │ 0.8538 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ daf204c           │ 05:02 PM │    2 │ 0.72421 │ 0.8284 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ bdb975c           │ 05:02 PM │    1 │  1.2537 │ 0.7738 │ 473987 │ 0.0001 │ 0            │
│ ├─╨ 77dc46c           │ 05:01 PM │    0 │  1.9428 │ 0.5715 │ 473987 │ 0.0001 │ 0            │
```

## Starting from an existing checkpoint

Since you have all of these different checkpoints, you might want to resume
training from a particular one. For example, maybe your accuracy started
decreasing at a certain checkpoint and you want to make some changes to fix
that.

First, we need to apply the checkpoint we want to begin our new experiment from.
To do that, run the following command:

```bash
dvc exp apply b3de55f
```

where _b3de55f_ is the id of the checkpoint you want to reference.

> Next, we'll change the learning rate in the _params.yaml_ to `0.001`. Then we'll

Reviewer: just an idea:
```
dvc exp run --set-param lr=0.001
``` 
might be an easier way (for user) of doing it.

Author: @dmpetrov @dberenbaum I tried running this command in few different ways and I get this error each time:

```
Traceback (most recent call last):
  File "/Users/milecia/Repos/checkpoints-tutorial/train.py", line 123, in <module>
    main()
  File "/Users/milecia/Repos/checkpoints-tutorial/train.py", line 110, in main
    train(model, x_batch, y_batch, params["lr"], params["weight_decay"])
  File "/Users/milecia/Repos/checkpoints-tutorial/train.py", line 45, in train
    optimizer = torch.optim.Adam(model.parameters(), lr=lr,
  File "/Users/milecia/Repos/checkpoints-tutorial/.venv/lib/python3.9/site-packages/torch/optim/adam.py", line 36, in __init__
    if not 0.0 <= lr:
TypeError: '<=' not supported between instances of 'float' and 'str'
```

Any thoughts or suggestions?

│ workspace             │ -            │   18 │ 0.12399  │ 0.9625 │ 473987 │ 0.001  │ 0            │
│ main                  │ Apr 27, 2021 │    - │       -  │      - │ 473987 │ 0.0001 │ 0            │
│ │ ╓ exp-63287         │ 04:37 PM     │   18 │ 0.12399  │ 0.9625 │ 473987 │ 0.001  │ 0            │
│ │ ╟ 868c144           │ 04:37 PM     │   17 │ 0.12042  │ 0.9616 │ 473987 │ 0.001  │ 0            │
│ │ ╟ 4de4e69           │ 04:37 PM     │   16 │ 0.13808  │ 0.9585 │ 473987 │ 0.001  │ 0            │
│ │ ╟ d8d633c           │ 04:37 PM     │   15 │ 0.20415  │ 0.9397 │ 473987 │ 0.001  │ 0            │
│ │ ╟ 8fbab93 (b3de55f) │ 04:36 PM     │   14 │ 0.24509  │ 0.9309 │ 473987 │ 0.001  │ 0            │
│ │ ╓ exp-4de2a         │ 05:03 PM     │    7 │  0.3096  │ 0.9092 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ d50c724           │ 05:03 PM     │    6 │ 0.33246  │ 0.9044 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ 29491a9           │ 05:02 PM     │    5 │ 0.36601  │ 0.8943 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ b3de55f           │ 05:02 PM     │    4 │ 0.41666  │ 0.8777 │ 473987 │ 0.0001 │ 0            │
```

> The existing checkpoint is referenced at the beginning of the new experiment.

Reviewer: The sha-s would be super helpful here. It might be not clear for the user what is existing and a new mean.

Author: What do you mean by the 'sha-s'?

Reviewer 2: @flippedcoder This means not using the `exp-4de2a` names but the Git commit hashes instead. I think if you upgrade to the latest dvc version, you will be able to see both and it will fix this issue.

Author: Gotcha. I think I have the latest DVC version. Is it `2.0.18`? I still don't see the Git hashes though. Just the `exp-f923ji` syntax.

Reviewer 2: It's 2.1.0 now. The table format hopefully looks something like:
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┓
┃ Experiment              ┃ Created      ┃ step ┃ acc    ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━┩
│ workspace               │ -            │ 1    │ 0.9133 │
│ test                    │ 12:52 PM     │ 1    │ 0.9133 │
│ live                    │ Apr 19, 2021 │ -    │ -      │
│ │ ╓ 7243788 [exp-bffd1] │ 12:46 PM     │ 1    │ 0.9133 │
│ ├─╨ c039cb0             │ 12:46 PM     │ 0    │ 0.8523 │
│ │ ╓ 9121bba [exp-dbb5a] │ 12:46 PM     │ 6    │ 0.9791 │
│ │ ╟ 9790bd2             │ 12:45 PM     │ 5    │ 0.9807 │
│ │ ╟ 5d629a2             │ 12:45 PM     │ 4    │ 0.976  │
│ │ ╟ b94ecb0             │ 12:45 PM     │ 3    │ 0.9714 │
│ │ ╟ 5a7ba19             │ 12:45 PM     │ 2    │ 0.9569 │
│ │ ╟ 0849314             │ 12:45 PM     │ 1    │ 0.9659 │
│ ├─╨ 98c13ed             │ 12:44 PM     │ 0    │ 0.95   │
└─────────────────────────┴──────────────┴──────┴────────┘
```

The new experiment is referred to as a modified experiment because it's taking
existing data and using it as the starting point.

### Metrics diff

When you've run all the experiments you want to and you are ready to compare
metrics between checkpoints, you can run the command:

```bash
dvc metrics diff 8fbab93 868c144
```

Make sure that you replace `8fbab93` and `868c144` with checkpoint ids from your
table with the checkpoints you want to compare. You'll see something similart to
this in your terminal.

```dvc
Path          Metric    Old      New      Change
dvclive.json  acc       0.9854   0.9831   -0.0023
dvclive.json  loss      0.04828  0.05095  0.00267
dvclive.json  step      19       15       -4

> ```

Reviewer: You can reference that these are the same numbers as you see in the table but in a different format.

with the following command:

```bash
dvc exp gc --workspace --force
```

## Adding checkpoints to Git

When you terminate training, you'll see a few commands in the terminal that will
allow you to add these changes to Git.

```dvc
To track the changes with git, run:

        git add dvclive.json dvc.yaml .gitignore train.py dvc.lock

Reproduced experiment(s): exp-263da
Experiment results have been applied to your workspace.

To promote an experiment to a Git branch run:

        dvc exp branch <exp>
```

You can run the following command to save your experiments to the Git history.

```bash
git add dvclive.json dvc.yaml .gitignore train.py dvc.lock
```

You can take a look at what will be commited to your Git history by running:

```bash
git status
```

You should see something similar to this in your terminal.

```git
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   .gitignore
        new file:   dvc.lock
        new file:   dvclive.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        data/
        dvclive.html
        dvclive/
        model.pt
        plots.html
        predictions.json

> ```

Reviewer: It looks like a commit is missing: 
```
git commit -m 'my experiment'
```

and, probably a last "motivational" sentence about checkpoint 😄

- On Windows: `.\.venv\Scripts\activate`

Once you have your environment set up, you can install the dependencies by
running:

```bash
pip install -r requirements.txt
```

This will download all of the packages you need to run the example. Now you have
everything you need to get started with experiments and checkpoints.

## Setting up a DVC pipeline

DVC versions data and models and it also can version the machine learning model
weights file, the checkpoints, during the training process. To enable this, you
will need to run your model training under DVC pipeline.

Adding DVC pipline only takes a few commands. At the root of the project, run:

```bash
dvc init
```

This sets up the files you need for your DVC pipeline to work.

This will create a _dvc.yaml_ file in your root directory.

Now we need to add a stage for training our model within a DVC pipeliene. We'll
do that with `dvc stage add`, which we'll explain more later. For now, run the
following command:

```bash
dvc stage add -n train -d data/MNIST -d train.py \
              -c model.pt --plots-no-cache predictions.json \
              -p seed,lr,weight_decay --live dvclive python train.py

> ```

Reviewer 2: Thoughts on using long flag names (`--name`) and using one flag per line to make this more readable? cc @dmpetrov @jorgeorpinel

Reviewer 3: Agree to use full option names and one arg per line at least the first time they're shown, but only the ones related to the topic, I think. I think it's OK to keep `-n train` along with the `-d ...` in the first line though.

The checkpoints need to be enabled in DVC at the pipeline level. The
`-c / --checkpoint` option of the `dvc stage add` command defines the checkpoint

> file or directory.

Reviewer 2: I'm not sure it's clear why `model.pt` is the checkpoint file here and not `predictions.json` or some other output. It might be helpful to clarify that the checkpoint file is an output from one checkpoint that becomes a dependency for the next checkpoint, such as the model weights file. cc @jorgeorpinel @pmrowla

Reviewer 3: Agree, introduce "circular dependency" concept (not sure about the terminology though)

The rest of the `dvc stage add` command sets up our dependencies for running the
training code, which parameters we want to track (which are defined in the
_params.yaml_), some configurations for our plots, showing the training metrics,
and specifying where the logs produced by the training process will go.

After running the command above to setup your _train_ stage, your _dvc.yaml_
should have the following code.

```yaml
stages:
  download:
    cmd: python download.py
    deps:
      - download.py
    outs:

>       - data/MNIST

Reviewer 2: Where did this stage come from? I don't see it in the repo or the above instructions. Is it from an old version?

Author: Oh yeah. That was from an old version.

relevant to checkpoints is when we iterate over the training epochs.

This is where DVC will be able to track our metrics over time and where we will
add our checkpoints to give us the points in time with our model that we can
switch between.

Now we need to enable checkpoints at the code level. We are interested in
tracking the metrics along with each checkpoint, so we'll need to add a few
lines of code.

In the `train.py` file, import the `dvclive` package with the other imports:

```bash
import dvclive
```

Then update the following lines of code in the `main` method inside of the
training epoch loop.

```python
# Iterate over training epochs.
for i in range(1, EPOCHS+1):
    # Train in batches.
    train_loader = torch.utils.data.DataLoader(
            dataset=list(zip(x_train, y_train)),
            batch_size=512,
            shuffle=True)
    for x_batch, y_batch in train_loader:
        train(model, x_batch, y_batch, params["lr"], params["weight_decay"])
    torch.save(model.state_dict(), "model.pt")
    # Evaluate and checkpoint.
    metrics = evaluate(model, x_test, y_test)
    for k, v in metrics.items():

>         print('Epoch %s: %s=%s'%(i, k, v))

Reviewer 2: It seems like this code block is missing the `dvclive` additions?

Reviewer 2: You could change the code format from `python` to `diff` and do:

```suggestion
-        print('Epoch %s: %s=%s'%(i, k, v))
+        dvclive.log(k, v)
+    dvclive.next_step()
```

```

The line `torch.save(model.state_dict(), "model.pt")` updates the checkpoint

> file.

Reviewer 2: This isn't an update to the existing code, though, right? I'm not sure this line is needed, or maybe it belongs somewhere else.

Author: Right, it's not an update to the code. It's just an explanation of where checkpoints are created in the code. I added it per this comment https://github.com/iterative/dvc.org/pull/2373#discussion_r622475955

## Running experiments

With checkpoints enabled and working in our code, let's run the experiment. You
can run an experiment with the following command:

```bash
dvc exp run
```

You'll see output similar to this in your terminal while the training process is
going on.

```dvc
Epoch 1: loss=1.9225023984909058
Epoch 1: acc=0.5833
Updating lock file 'dvc.lock'
Checkpoint experiment iteration '09aa592'.

file:///Users/milecia/checkpoints-tutorial/dvclive.html
Epoch 2: loss=1.2303115129470825
Epoch 2: acc=0.7777
Updating lock file 'dvc.lock'
Checkpoint experiment iteration 'f5742b3'.

file:///Users/milecia/checkpoints-tutorial/dvclive.html
Epoch 3: loss=0.714752733707428
Epoch 3: acc=0.8295
Updating lock file 'dvc.lock'
Checkpoint experiment iteration '4917b0e'.

file:///Users/milecia/checkpoints-tutorial/dvclive.html
Epoch 4: loss=0.5060071349143982
Epoch 4: acc=0.8551
Updating lock file 'dvc.lock'
Checkpoint experiment iteration 'acfa695'.

file:///Users/milecia/checkpoints-tutorial/dvclive.html
Epoch 5: loss=0.41547226905822754
Epoch 5: acc=0.8785
Updating lock file 'dvc.lock'
Checkpoint experiment iteration '69beaaa'.
```

> After a few epochs have completed, stop terminate the training process with

Reviewer 2: ```suggestion
After a few epochs have completed, stop the training process with
```

command:

```bash
dvc exp run --set-param lr=0.001
```

You'll be able to see where the experiment starts from the existing checkpoint
by running:

```bash
dvc exp show
```

You should seem something similar to this in your terminal.

```dvc
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━┓
┃ Experiment            ┃ Created      ┃ step ┃     loss ┃    acc ┃ seed   ┃ lr     ┃ weight_decay ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━┩
│ workspace             │ -            │   18 │ 0.12399  │ 0.9625 │ 473987 │ 0.001  │ 0            │
│ main                  │ Apr 27, 2021 │    - │       -  │      - │ 473987 │ 0.0001 │ 0            │
│ │ ╓ exp-63287         │ 04:37 PM     │   18 │ 0.12399  │ 0.9625 │ 473987 │ 0.001  │ 0            │
│ │ ╟ 868c144           │ 04:37 PM     │   17 │ 0.12042  │ 0.9616 │ 473987 │ 0.001  │ 0            │
│ │ ╟ 4de4e69           │ 04:37 PM     │   16 │ 0.13808  │ 0.9585 │ 473987 │ 0.001  │ 0            │
│ │ ╟ d8d633c           │ 04:37 PM     │   15 │ 0.20415  │ 0.9397 │ 473987 │ 0.001  │ 0            │

> │ │ ╟ 8fbab93 (b3de55f) │ 04:36 PM     │   14 │ 0.24509  │ 0.9309 │ 473987 │ 0.001  │ 0            │

Reviewer 2: The `step` value here should be 5 based on `b3de55f`. Can you try with the most recent version of dvc and dvclive? If you still have this issue, it might be a bug in dvc or dvclive. cc @pmrowla @pared

Reviewer 2: It also doesn't seem like the model weights were applied correctly here since it's unlikely they jumped from 0.8777 in b3de55f to 0.9309 in 8fbab93.

Reviewer 4: @flippedcoder Did you try with the latest version? Is the problem still occuring?

metrics between checkpoints, you can run the command:

```bash
dvc metrics diff 8fbab93 868c144
```

Make sure that you replace `8fbab93` and `868c144` with checkpoint ids from your
table with the checkpoints you want to compare. You'll see something similar to
this in your terminal.

```dvc
Path          Metric    Old      New      Change
dvclive.json  acc       0.9854   0.9831   -0.0023
dvclive.json  loss      0.04828  0.05095  0.00267
dvclive.json  step      19       15       -4
```

_These are the same numbers you see in the metrics table, just in a different
format._

### Looking at plots

You also have the option to generate plots to visualize the metrics about your
training epochs. Running:

```bash
dvc plots diff 8fbab93 868c144
```

where `8fbab93` and `868c144` are the checkpoint ids you want to compare, will
generate a `plots.html` file that you can open in a browser and it will display
plots for you similar to the ones below.

![Plots generated from running experiments on MNIST dataset using DVC](/img/checkpoints_plots.png)
_The results here won't show anything interesting because the model increases in

> accuracy so fast._

Reviewer 2: It seems like this plot is not only uninteresting, but also kind of confusing since there's only one line visible on it despite comparing two checkpoints. If you compare the last checkpoints in both the original and the modified experiment, you should end up with something more interesting since the different learning rates should cause them to diverge.

│ workspace             │ -        │    4 │ 0.41666 │ 0.8777 │ 473987 │ 0.0001 │ 0            │
│ main                  │ 04:49 PM │    - │       - │      - │ 473987 │ 0.001  │ 0            │
│ │ ╓ exp-a75bb         │ 05:18 PM │    4 │ 0.41666 │ 0.8777 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ b8a4ceb           │ 05:18 PM │    3 │ 0.50835 │ 0.8538 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ 7b652b1           │ 05:17 PM │    2 │ 0.72421 │ 0.8284 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ cf0168f           │ 05:17 PM │    1 │  1.2537 │ 0.7738 │ 473987 │ 0.0001 │ 0            │
│ ├─╨ 8c7f40f           │ 05:17 PM │    0 │  1.9428 │ 0.5715 │ 473987 │ 0.0001 │ 0            │
│ │ ╓ exp-7fafd         │ 05:10 PM │    5 │ 0.36601 │ 0.8943 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ 9ac9e6a           │ 05:09 PM │    4 │ 0.41666 │ 0.8777 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ b869efe           │ 05:09 PM │    3 │ 0.50835 │ 0.8538 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ 422395e           │ 05:09 PM │    2 │ 0.72421 │ 0.8284 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ 44e44a9           │ 05:09 PM │    1 │  1.2537 │ 0.7738 │ 473987 │ 0.0001 │ 0            │

> │ ├─╨ c6e11b4           │ 05:09 PM │    0 │  1.9428 │ 0.5715 │ 473987 │ 0.0001 │ 0            │

Reviewer 2: Seems like this example is from a different set of experiments. It probably makes sense to clean these up and reproduce all the experiment results once everything else is pretty much resolved.

```

We can also remove all of the experiments we don't promote to our Git workspace
with the following command:

```bash
dvc exp gc --workspace --force

> ```

Reviewer 2: I'd suggest moving this to the bottom of the next section since it won't do anything if the users run it at this point before committing any experiment results to Git.

## Adding checkpoints to Git

When you terminate training, you'll see a few commands in the terminal that will
allow you to add these changes to Git.

```dvc
To track the changes with git, run:

        git add dvclive.json dvc.yaml .gitignore train.py dvc.lock

Reproduced experiment(s): exp-263da
Experiment results have been applied to your workspace.

To promote an experiment to a Git branch run:

>         dvc exp branch <exp>

Reviewer 2: If you want to keep all of the checkpoints as commits in your Git history, you can use `dvc exp branch` to do that. I'm torn on whether it's worthwhile to explain that here. cc @dmpetrov @jorgeorpinel

Reviewer 3: I wouldn't mention `exp branch` out of context here. Actually I'd skip the whole code block above, as the `git` command is repeated below.

    train_loader = torch.utils.data.DataLoader(
            dataset=list(zip(x_train, y_train)),
            batch_size=512,
            shuffle=True)
    for x_batch, y_batch in train_loader:
        train(model, x_batch, y_batch, params["lr"], params["weight_decay"])
    torch.save(model.state_dict(), "model.pt")
    # Evaluate and checkpoint.
    metrics = evaluate(model, x_test, y_test)
    for k, v in metrics.items():
-        print('Epoch %s: %s=%s'%(i, k, v))
+        dvclive.log(k, v)
+    dvclive.next_step()
```

The line `torch.save(model.state_dict(), "model.pt")` updates the checkpoint
file.

The `dvclive.log(k, v)` line stores the metric _k_ with a value _v_ in plain
text files in the _dvclive_ directory by default.

The `dvclive.next_step()` line tells DVC that it can take a snapshot of the
entire workspace and version it with Git. It's important that with this approach
only code with metadata is versioned in Git (as an ephemeral commit), while the
actual model weight file will be stored in the DVC data cache.

## Running experiments

With checkpoints enabled and working in our code, let's run the experiment. You
can run an experiment with the following command:

> ```bash
> dvc exp run

Reviewer 3: Please use the `dvc` highlighter (instead of `bash`) and `$` prompts throughout. Here's one:
````suggestion
```dvc
$ dvc exp run
````

```

You'll see output similar to this in your terminal while the training process is
going on.

> ```dvc

Reviewer 3: In its current form (no commands) this one doesn't need highlighting:
````suggestion
```
````

│ │ ╟ 422395e           │ 05:09 PM │    2 │ 0.72421 │ 0.8284 │ 473987 │ 0.0001 │ 0            │
│ │ ╟ 44e44a9           │ 05:09 PM │    1 │  1.2537 │ 0.7738 │ 473987 │ 0.0001 │ 0            │
│ ├─╨ c6e11b4           │ 05:09 PM │    0 │  1.9428 │ 0.5715 │ 473987 │ 0.0001 │ 0            │
```

We can also remove all of the experiments we don't promote to our Git workspace
with the following command:

```bash
dvc exp gc --workspace --force
```

## Adding checkpoints to Git

When you terminate training, you'll see a few commands in the terminal that will
allow you to add these changes to Git.

```dvc
To track the changes with git, run:

        git add dvclive.json dvc.yaml .gitignore train.py dvc.lock

Reproduced experiment(s): exp-263da
Experiment results have been applied to your workspace.

To promote an experiment to a Git branch run:

        dvc exp branch <exp>
```

You can run the following command to save your experiments to the Git history.

```bash
git add dvclive.json dvc.yaml .gitignore train.py dvc.lock
```

You can take a look at what will be commited to your Git history by running:

```bash
git status
```

You should see something similar to this in your terminal.

> ```git

Reviewer 3: Same for this one and any other that's just terminal output:
````suggestion
```
````

# Checkpoints

ML checkpoints are an important part of deep learning because ML engineers like
to save the model files at certain points during a training process.

With DVC experiments and checkpoints, you can:

- Implement the best practice in deep learning to save your model weights as
  checkpoints.
- Track all code and data changes corresponded to the checkpoints.
- See when metrics start diverging and revert to the optimal checkpoint.
- Automate the process of tracking every training epoch.

[The way checkpoints are implemented by DVC](/blog/experiment-refs) utilizes
_ephemeral_ experiment commits and experiment branches within DVC. They are
created using the metadata from experiments and are tracked with the `exps`
custom Git reference.

You can add experiments to your Git history by committing the experiment you

> want to track, which you'll see later in this t$ utorial.

Reviewer 2: ```suggestion
want to track, which you'll see later in this tutorial.
```
